TestUtilities/Sources/TestUtilities/Helpers/DDAssert.swift

```diff
@@ -0,0 +1,209 @@
+        message = keyPath.isEmpty ? message : message + " at keyPath " + keyPath.joined(separator: ".")
+        XCTFail(message, file: file, line: line)
+    } catch {
+        let message = message() + " - threw error \"\(error)\""
+        XCTFail(message, file: file, line: line)
+    }
+}
+
+private func _DDAssertReflect(_ expression1: @autoclosure () throws -> Any, _ expression2: @autoclosure () throws -> Any, keyPath: [String] = []) throws {
+    let (value1, value2) = (try expression1(), try expression2())
+    let mirror1 = Mirror(reflecting: value1)
+    let mirror2 = Mirror(reflecting: value2)
+
+    guard mirror1.displayStyle == mirror1.displayStyle else {
+        throw DDAssertError.expectedFailure("(\"\(value1)\") and (\"\(value2)\") have different types", keyPath: keyPath)
+    }
+
+    guard mirror1.children.count == mirror1.children.count else {
+        throw DDAssertError.expectedFailure("(\"\(value1)\") and (\"\(value2)\") have different number of children", keyPath: keyPath)
+    }
+
+    if mirror1.children.isEmpty, mirror1.children.isEmpty {
+        guard String(describing: value1) == String(describing: value2) else { // plain values, compare debug strings
+            throw DDAssertError.expectedFailure("(\"\(value1)\") is not equal to (\"\(value2)\")", keyPath: keyPath)
+        }
+
+        return
+    }
+
+    switch (mirror1.displayStyle, mirror2.displayStyle) {
+    case (.dictionary?, .dictionary?): // two dictionaries
+        let dictionary1 = value1 as! [String: Any]
+        let dictionary2 = value2 as! [String: Any]
+
+        guard dictionary1.keys.count == dictionary2.keys.count else {
+            throw DDAssertError.expectedFailure("dictionaries have different number of keys", keyPath: keyPath)
+        }
+
+        for (key1, value1) in dictionary1 {
+            guard let value2 = dictionary2[key1] else {
+                throw DDAssertError.expectedFailure("dictionaries have different key names", keyPath: keyPath)
+            }
+
+            try _DDAssertReflect(value1, value2, keyPath: keyPath + [key1])
+        }
+
+        return // dictionaries are equal
+    case (.set?, .set?): // two sets
+        let set1 = value1 as! Set<AnyHashable>
+        let set2 = value2 as! Set<AnyHashable>
+
+        guard set1 == set2 else {
+            throw DDAssertError.expectedFailure("sets are not equal", keyPath: keyPath)
+        }
+
+        return // sets are equal
+    default:
+        break // other than dictionary or set, continue...
+    }
+
+    for (child1, child2) in zip(mirror1.children, mirror2.children) { // compare each child
+        let key = child1.label.map { [$0] } ?? []
+        try _DDAssertReflect(child1.value, child2.value, keyPath: keyPath + key)
+    }
+}
+
+public func DDAssertReflect<T>(_ expression1: @autoclosure () throws -> T, _ expression2: @autoclosure () throws -> T, _ message: @autoclosure () -> String = "", file: StaticString = #filePath, line: UInt = #line) {
```

Comment:
nit/ The name could be quite vague, hence adding a comment could be priceless for newcomers.

```suggestion
/// Asserts that both expressions equal by using reflection and comparing their mirrors.
public func DDAssertReflect<T>(_ expression1: @autoclosure () throws -> T, _ expression2: @autoclosure () throws -> T, _ message: @autoclosure () -> String = "", file: StaticString = #filePath, line: UInt = #line) {
```